src/pointer-analysis/value_set.h

```diff
@@ -298,30 +494,81 @@ class value_sett
     get_reference_set_rec(expr, dest, ns);
   }
 
+  /// See get_reference_set.
   void get_reference_set_rec(
     const exprt &expr,
     object_mapt &dest,
     const namespacet &ns) const;
 
+  /// Sets `dest` to `src` with any wrapping typecasts removed
   void dereference_rec(
     const exprt &src,
     exprt &dest) const;
 
-  void assign_rec(
-    const exprt &lhs,
-    const object_mapt &values_rhs,
-    const std::string &suffix,
-    const namespacet &ns,
-    bool add_to_sets);
-
+  /// Marks objects that may be pointed to by `op` as maybe-invalid
+  /// \param op: pointer to invalidate
+  /// \param ns: global namespace
   void do_free(
     const exprt &op,
     const namespacet &ns);
 
+  /// Extracts a member from a struct- or union-typed expression.
+  /// Usually that means making a `member_exprt`, but this can shortcut
+  /// extracting members from constants or with-expressions.
+  /// \param src: base struct-or-union-typed expression
+  /// \param component_name: member to extract
+  /// \param ns: global namespace
   exprt make_member(
     const exprt &src,
     const irep_idt &component_name,
     const namespacet &ns);
+
+  // Subclass customisation points:
+
+protected:
+  /// Subclass customisation point for recursion over RHS expression:
+  virtual void get_value_set_rec(
+    const exprt &expr,
+    object_mapt &dest,
+    const std::string &suffix,
+    const typet &original_type,
+    const namespacet &ns) const;
+
+  /// Subclass customisation point for recursion over LHS expression:
+  virtual void assign_rec(
+    const exprt &lhs,
+    const object_mapt &values_rhs,
+    const std::string &suffix,
+    const namespacet &ns,
+    bool add_to_sets);
+
+  /// Subclass customisation point for applying code to this domain:
+  virtual void apply_code_rec(
+    const codet &code,
+    const namespacet &ns);
+
+ private:
+  /// Subclass customisation point to filter or otherwise alter the value-set
+  /// returned from get_value_set before it is passed into assign. For example,
+  /// this is used in one subclass to tag and thus differentiate values that
+  /// originated in a particular place, vs. those that have been copied.
+  virtual void adjust_assign_rhs_values(
```

Comment:
Why is this marked  `private` (yet `virtual`)?!

Reply:
There's no base implementation for the subclass implementation to call, so no need to make it protected. Could make it so if you'd prefer subclasses to always call it in case it grows one, but any such base class behaviour could be directly implemented at the callsite.